### ISSUE TITLE
Minimal Changes for Batch based Inference

### ISSUE THREAD

**Issue by dcyoung** [armandmcqueen/tensorpack-mask-rcnn#2]
I've been following the work in this repo and am excited to see progress on batch based training. 

I encountered a few issues attempting to run your code on my local machines, though I understand this is a very WIP repo.

I was wondering if you had experimented with batch based inference? I'm hoping you might be willing to offer some advice on a minimal set of changes to support inference on > 1 image. Perhaps something less invasive than the full set of changes required for training, but not sure if that is possible.

Thanks.

**Comment by ppwwyyxx** [comment]
Found some great work going on here!

The massive changes here do look a bit intimidating. I suppose what's needed as a starting point is mainly:
1. split / concat on the batch dimension for critical paths (backbone & head network)
2. python-side for loop over batch dimension, for the other parts

I imagine this won't require a huge change and can give a reasonable speedup already, especially if batching is done with aspect-ratio awareness. Then the python-side loops can be further optimized one by one.

**Comment by armandmcqueen** [comment]
Hi @dcyoung,

Apologies for leaving this question unanswered for so long. We only recently got to the point where the code is working enough to use.

We have focused heavily on getting training to run successfully with per GPU batch size > 1. The training Mask RCNN graph now works with batches of images.  For inference, you reuse the same graph, with some slight changes near the end (outputting bounding boxes instead of computing losses, etc.). This means that the majority of the work for batch inference is done. We are still slightly off in training accuracy for a few configurations, which can indicate a bug in the graph (which would affect inference) or a bug in loss/hyperparams (which would not), but we are actively working on addressing that. 

We have not yet spent a lot of time working on the parts of the codebase that are not related to training. Some of these would need to be brought up to date to do batch inference: 

(1) In `model/`, which contains the code for the core graph, you would need to look through the code and confirm that that non-training codepaths are working correctly. Those codepaths look like this:

```
if training: 
# sometimes 'if self.training:'
  # code that works with batch
else:
  # code you would potentially need to change
```

This is probably not a huge amount of work as the non-training codepaths are less complicated than the training codepaths.

(2) Then you would need to fix the inference entrypoint - in `train.py`, you would likely want to use `do_predict`. This would need to be changed to work with batches - confirm that the inference tensor names are still correct, figure out how to take in multiple images from the command line, pad and combine them*  and then rewrite `predict_image` to handle batches (there is a `predict_image_batch` function, but we haven't looked at it in a while so I'm not sure if it works)

\* Look to `data.py:get_batch_train_dataflow()` as an example of how to batchify image inputs. It is essentially padding the images so they all have the same dimensions, concating them along a batch dimension and then passing the original image dimensions into the graph so that the graph can figure out how to ignore the padding.

---

One we have debugged all the training issues, I will come back to making sure that inference works correctly.

**Comment by dcyoung** [comment]
Thank you both for the input. For my use-case, a uniform image size is a given, so this may not be too much work. I'll investigate an implementation based on this feedback. Thanks
